[RFC] gpio-cdev: add nu801 userspace driver

riptidewave93 opened the pull request:
This adds a userspace interpretation of the nu801 driver used by Meraki
hardware. Previously this was a driver that was added per target, but as
multiple targets now have this driver, we should move to something that
can be shared by all targets since no driver exists upstream.

Co-developed-by: Christian Lamparter <chunkeey@gmail.com>
Signed-off-by: Chris Blake <chrisrblake93@gmail.com>

So for additional details, the NU801 used to be used by the MR18 and Z1
boards under the old ath79 target using the Meraki GPL driver. This code can
be found at https://github.com/openwrt/openwrt/blob/v19.07.7/target/linux/ar71xx/files/drivers/leds/leds-nu801.

Since this driver was never upstreamed, and since multiple targets use this hardware,
we should probably move to using a cdev based driver that runs in userspace. This
allows it to be packaged per target, instead of throwing the nu801s kernel driver
in generic.

Note that there are some questions still since cdev based LED drivers are not yet
"supported" in OpenWRT. For example, is this a good way to start the service? How
can we get sysupgrade to support this? Is this a preferred implementation over nu801
being thrown in as a kernel patch under generic? Should cdev support be enabled in
generic?

The only downside to this method is it will require the 5.10 kernel since 
CONFIG_GPIO_CDEV was added in kernel 5.10.

Tested and working on a Meraki MX100 (slowly working on upstreaming into OpenWRT)

I am open to feedback/input/PRs to my source branch. Thanks for the consideration.

adschm commented on package/system/gpio-cdev/nu801/Makefile:
```diff
@@ -0,0 +1,46 @@
+# Copyright (C) 2021 OpenWrt.org
+#
+# This is free software, licensed under the GNU General Public License v2.
+# See /LICENSE for more information.
```
Please use SPDX identifier instead.

adschm commented on package/system/gpio-cdev/nu801/files/nu801.init:
```diff
@@ -0,0 +1,14 @@
+	procd_set_param command /usr/sbin/nu801 "${BOARD_MODEL}"
+	procd_set_param respawn 5 1 5
+	procd_close_instance
+}
```
Missing newline at EOF.

adschm commented on package/system/gpio-cdev/nu801/files/nu801.init:
```diff
@@ -0,0 +1,14 @@
+#!/bin/sh /etc/rc.common
+# Copyright (C) 2021 OpenWrt.org
```
Please add an SPDX identifier.

adschm commented on package/system/gpio-cdev/nu801/Makefile:
```diff
@@ -0,0 +1,46 @@
+#
+# Copyright (C) 2021 OpenWrt.org
```
Unless this is put there by Christian, note that you are not OpenWrt, so you may not put this here.

chunkeey commented:
@linusw 

> The only downside to this method is it will require the 5.10 kernel since
> CONFIG_GPIO_CDEV was added in kernel 5.10.

Since this is GPIO_CDEV (V2). Do you reckon that V2 of this API will stick around?

riptidewave93 committed:
gpio-cdev: add nu801 userspace driver

This adds a userspace interpretation of the nu801 driver used by Meraki
hardware. Previously this was a driver that was added per target, but as
multiple targets now have this driver, we should move to something that
can be shared by all targets since no driver exists upstream.

Co-developed-by: Christian Lamparter <chunkeey@gmail.com>
Signed-off-by: Chris Blake <chrisrblake93@gmail.com>

riptidewave93 commented:
@adschm I have addressed your feedback so far, thank you for the input!

For anyone who has a moment id love to get some of the questions I asked in the PR answered, mainly:
1. Is this the right approach, or should I instead work on getting the nu801 kernel driver into generic?
2. Are there any reservations of this method, mainly around the fact it will REQUIRE kernel 5.10 or newer?
3. If cdev is the path forward, what other implementation work may be needed? sysupgrade support? better integration? etc.

riptidewave93 commented:
Going to close for now, will bring back later after some additional work is done for better integration.
